dex/networks/eth/contracts/ETHSwapV1.sol

```diff
@@ -0,0 +1,221 @@
+
+    // state returns the current state of the swap.
+    function state(Contract calldata c)
+        public view returns(State)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveRecord(c);
+
+        if (blockNum == 0) {
+            return State.Empty;
+        }
+        if (record == RefundRecord) {
+            return State.Refunded;
+        }
+        if (secretValidates(record, c.secretHash)) {
+            return State.Redeemed;
+        }
+        // Is it worth checking whether blockNum < block.number?
+        return State.Filled;
+    }
+
+    // initiate initiates an array of Contracts.
+    function initiate(Contract[] calldata contracts)
+        public
+        payable
+        senderIsOrigin()
+    {
+        uint initVal = 0;
+        for (uint i = 0; i < contracts.length; i++) {
+            Contract calldata c = contracts[i];
+
+            require(c.value > 0, "0 val");
+            require(c.refundTimestamp > 0, "0 refundTimestamp");
+
+            // Is this needed?
+            // require(msg.sender == c.initiator)
+
+            bytes32 k = contractKey(c);
+            bytes32 record = swaps[k];
+            require(record == bytes32(0), "swap not empty");
+
+            swaps[k] = bytes32(block.number);
+
+            initVal += c.value * 1 gwei;
+        }
+
+        require(initVal == msg.value, "bad val");
+    }
+
+    // isRedeemable returns whether or not a swap identified by secretHash
+    // can be redeemed using secret.
+    function isRedeemable(Contract calldata c)
+        public
+        view
+        returns (bool)
+    {
+        (, bytes32 record, uint256 blockNum) = retrieveRecord(c);
+        return blockNum != 0 && blockNum <= block.number && !secretValidates(record, c.secretHash);
+    }
+
+    // redeem redeems a Contract. It checks that the sender is not a contract,
+    // and that the secret hash hashes to secretHash. msg.value is tranfered
+    // from ETHSwap to the sender.
+    //
+    // To prevent reentry attack, it is very important to check the state of the
+    // contract first, and change the state before proceeding to send. That way,
+    // the nested attacking function will throw upon trying to call redeem a
+    // second time. Currently, reentry is also not possible because contracts
+    // cannot use this contract.
+    function redeem(Redemption[] calldata redemptions)
+        public
+        senderIsOrigin()
+    {
+        uint amountToRedeem = 0;
+        for (uint i = 0; i < redemptions.length; i++) {
+            Redemption calldata r = redemptions[i];
+
+            require(r.c.participant == msg.sender, "not authed");
+
+            (bytes32 k, bytes32 record, uint256 blockNum) = retrieveRecord(r.c);
+
+            // To be redeemable, the record needs to represent a valid block
+            // number.
+            require(blockNum > 0 && blockNum < block.number, "unfilled swap");
+            // require(blockNum >> 32 == 0, "invalid swap format");
```

Comment:
Commented out, but does ethereum have this limitation of a uint32 block number?

Reply:
No, but it'll take a couple hundred years to get to that block height. This comment is left over from an earlier iteration on the record scheme, but `blockNum < block.number` is better.